File: brick_simulation_py/src/components/level.py
"""
Level component - manages brick layouts for different stages
"""
import random
from ..utils.helpers import calculate_brick_position

class Level:
    """
    Level class manages the brick layout and difficulty for each game stage
    """
    def __init__(self, config, level_number=1):
        self.config = config
        self.level_number = level_number
        self.bricks = []
    
    def create_bricks(self, brick_class):
        """
        Create bricks based on the level number
        Returns a list of Brick objects
        """
        self.bricks = []
        
        # Choose a pattern based on level number
        if self.level_number % 5 == 1:
            self._create_standard_pattern(brick_class)
        elif self.level_number % 5 == 2:
            self._create_pyramid_pattern(brick_class)
        elif self.level_number % 5 == 3:
            self._create_checkerboard_pattern(brick_class)
        elif self.level_number % 5 == 4:
            self._create_mixed_pattern(brick_class)
        else:  # level % 5 == 0
            self._create_fortress_pattern(brick_class)
            
        return self.bricks
    
    def _create_standard_pattern(self, brick_class):
        """Create a standard rectangular pattern of bricks"""
        rows = min(self.config.BRICK_ROWS, 5 + self.level_number // 3)
        
        for row in range(rows):
            for col in range(self.config.BRICK_COLS):
                x, y = calculate_brick_position(row, col, self.config)
                
                # Determine brick type based on row and level
                if self.level_number > 3 and row == 0:
                    brick_type = "strong"
                    color = self.config.BRICK_COLORS["strong"]
                    points = 20
                else:
                    brick_type = "normal"
                    # Different colors for different rows
                    color_values = [(255, 0, 0), (255, 165, 0), (255, 255, 0), 
                                   (0, 255, 0), (0, 0, 255)]
                    color = color_values[row % len(color_values)]
                    points = 10
                
                brick = brick_class(x, y, self.config.BRICK_WIDTH, self.config.BRICK_HEIGHT, 
                                  color, brick_type, points)
                self.bricks.append(brick)
    
    def _create_pyramid_pattern(self, brick_class):
        """Create a pyramid-like pattern of bricks"""
        max_cols = self.config.BRICK_COLS
        
        for row in range(6):
            # Calculate how many bricks in this row and their starting column
            bricks_in_row = max_cols - row * 2
            if bricks_in_row <= 0:
                break
                
            start_col = (max_cols - bricks_in_row) // 2
            
            for col in range(start_col, start_col + bricks_in_row):
                x, y = calculate_brick_position(row, col, self.config)
                
                # Special bricks at the top of the pyramid
                if row == 0 and (col == start_col or col == start_col + bricks_in_row - 1):
                    brick_type = "strong"
                    color = self.config.BRICK_COLORS["strong"]
                    points = 20
                else:
                    brick_type = "normal"
                    # Gradient colors from red to blue
                    red = max(0, 255 - row * 50)
                    blue = min(255, row * 50)
                    color = (red, 0, blue)
                    points = 10
                
                brick = brick_class(x, y, self.config.BRICK_WIDTH, self.config.BRICK_HEIGHT, 
                                  color, brick_type, points)
                self.bricks.append(brick)
    
    def _create_checkerboard_pattern(self, brick_class):
        """Create a checkerboard pattern of bricks"""
        rows = min(8, 5 + self.level_number // 2)
        
        for row in range(rows):
            for col in range(self.config.BRICK_COLS):
                # Skip every other brick in a checkerboard pattern
                if (row + col) % 2 == 0:
                    continue
                    
                x, y = calculate_brick_position(row, col, self.config)
                
                # Add some strong bricks in higher levels
                if self.level_number > 5 and random.random() < 0.2:
                    brick_type = "strong"
                    color = self.config.BRICK_COLORS["strong"]
                    points = 20
                else:
                    brick_type = "normal"
                    # Alternate colors for visual effect
                    colors = [(0, 128, 255), (255, 0, 128)]
                    color = colors[(row + col) % len(colors)]
                    points = 10
                
                brick = brick_class(x, y, self.config.BRICK_WIDTH, self.config.BRICK_HEIGHT, 
                                  color, brick_type, points)
                self.bricks.append(brick)
    
    def _create_mixed_pattern(self, brick_class):
        """Create a mixed pattern with some unbreakable bricks"""
        rows = 6
        special_layout = [
            "NNNNNNNNNN",
            "N--N--N--N",
            "NNNNNNNNNN",
            "N-NN--NN-N",
            "N--NNNN--N",
            "NNNNNNNNNN"
        ]
        
        # Add unbreakable bricks for higher levels
        if self.level_number > 6:
            special_layout[1] = "N-UN-UN-UN"
            
        for row, layout_row in enumerate(special_layout[:rows]):
            for col, cell in enumerate(layout_row[:self.config.BRICK_COLS]):
                if cell == "-":
                    continue  # Skip this position
                    
                x, y = calculate_brick_position(row, col, self.config)
                
                if cell == "U":
                    brick_type = "unbreakable"
                    color = self.config.BRICK_COLORS["unbreakable"]
                    points = 0
                elif cell == "S" or (self.level_number > 8 and random.random() < 0.3):
                    brick_type = "strong"
                    color = self.config.BRICK_COLORS["strong"]
                    points = 20
                else:
                    brick_type = "normal"
                    # Color based on position
                    hue = (row * 30 + col * 20) % 360
                    # Simple HSV to RGB conversion for variety
                    if hue < 60:
                        r, g, b = 255, int(255 * hue / 60), 0
                    elif hue < 120:
                        r, g, b = int(255 * (120 - hue) / 60), 255, 0
                    elif hue < 180:
                        r, g, b = 0, 255, int(255 * (hue - 120) / 60)
                    elif hue < 240:
                        r, g, b = 0, int(255 * (240 - hue) / 60), 255
                    elif hue < 300:
                        r, g, b = int(255 * (hue - 240) / 60), 0, 255
                    else:
                        r, g, b = 255, 0, int(255 * (360 - hue) / 60)
                    color = (r, g, b)
                    points = 10
                
                brick = brick_class(x, y, self.config.BRICK_WIDTH, self.config.BRICK_HEIGHT, 
                                  color, brick_type, points)
                self.bricks.append(brick)
    
    def _create_fortress_pattern(self, brick_class):
        """Create a fortress-like pattern with unbreakable outer wall"""
        rows = 8
        cols = self.config.BRICK_COLS
        
        for row in range(rows):
            for col in range(cols):
                # Create an outer wall of unbreakable bricks
                if row == 0 or row == rows-1 or col == 0 or col == cols-1:
                    brick_type = "unbreakable"
                    color = self.config.BRICK_COLORS["unbreakable"]
                    points = 0
                # Random strong bricks inside
                elif random.random() < 0.3:
                    brick_type = "strong"
                    color = self.config.BRICK_COLORS["strong"]
                    points = 20
                else:
                    brick_type = "normal"
                    color = self.config.BRICK_COLORS["normal"]
                    points = 10
                
                x, y = calculate_brick_position(row, col, self.config)
                brick = brick_class(x, y, self.config.BRICK_WIDTH, self.config.BRICK_HEIGHT, 
                                  color, brick_type, points)
                self.bricks.append(brick)
    
    def get_breakable_brick_count(self):
        """Count how many breakable bricks are in the level"""
        return sum(1 for brick in self.bricks 
                  if brick.brick_type != "unbreakable" and brick.is_active())
    
    def increase_difficulty(self):
        """Increase level difficulty by incrementing level number"""
        self.level_number += 1

File: brick_sim_js/js/components/level.js
/**
 * Brick Simulation Game - Level Component
 * Handles level generation, brick layouts, and progression
 */

class Level {
    /**
     * Creates a new level
     * @param {number} levelNumber - The level number
     * @param {number} screenWidth - The width of the game screen
     * @param {number} screenHeight - The height of the game screen
     */
    constructor(levelNumber, screenWidth, screenHeight) {
        this.levelNumber = levelNumber;
        this.screenWidth = screenWidth;
        this.screenHeight = screenHeight;
        this.bricks = [];
        this.brickRows = CONFIG.BRICK_ROWS + Math.min(Math.floor(levelNumber / 2), 3); // Increase rows with level
        this.brickCols = CONFIG.BRICK_COLS;
        this.brickWidth = CONFIG.BRICK_WIDTH;
        this.brickHeight = CONFIG.BRICK_HEIGHT;
        this.brickMargin = CONFIG.BRICK_MARGIN;
        this.topMargin = CONFIG.BRICK_TOP_MARGIN;
        this.totalBricks = 0;
        this.brokeBricks = 0;
        this.bricksWithItems = 0;
        
        // Initialize level
        this.generateLevel();
    }
    
    /**
     * Generates the level layout with bricks
     */
    generateLevel() {
        this.bricks = [];
        this.brokeBricks = 0;
        
        // Calculate total width needed for bricks
        const totalBrickWidth = this.brickCols * (this.brickWidth + this.brickMargin) - this.brickMargin;
        // Calculate the left margin to center the bricks horizontally
        const leftMargin = (this.screenWidth - totalBrickWidth) / 2;
        
        // Track total number of breakable bricks
        let breakableBricks = 0;
        
        // Generate brick layout based on level
        for (let row = 0; row < this.brickRows; row++) {
            for (let col = 0; col < this.brickCols; col++) {
                // Skip some bricks for interesting patterns based on level
                if (this.shouldSkipBrick(row, col)) {
                    continue;
                }
                
                // Calculate brick position
                const x = leftMargin + col * (this.brickWidth + this.brickMargin);
                const y = this.topMargin + row * (this.brickHeight + this.brickMargin);
                
                // Determine brick type based on level and position
                const type = this.getBrickType(row, col);
                
                // Create the brick
                const brick = new Brick(
                    x, y, 
                    this.brickWidth, this.brickHeight, 
                    type, row, col
                );
                
                // Count breakable bricks
                if (type.name !== 'unbreakable') {
                    breakableBricks++;
                    
                    // Determine if this brick should have an item
                    if (this.shouldHaveItem(row, col)) {
                        const itemType = this.getRandomItemType();
                        brick.setItem(itemType);
                        this.bricksWithItems++;
                    }
                }
                
                this.bricks.push(brick);
            }
        }
        
        this.totalBricks = breakableBricks;
    }
    
    /**
     * Determines if a brick should be skipped based on level patterns
     * @param {number} row - The row index
     * @param {number} col - The column index
     * @returns {boolean} True if the brick should be skipped, false otherwise
     */
    shouldSkipBrick(row, col) {
        // Level 1: Simple full pattern
        if (this.levelNumber === 1) {
            return false;
        }
        
        // Level 2: Checkerboard pattern
        if (this.levelNumber === 2) {
            return (row + col) % 2 === 1;
        }
        
        // Level 3: Nested rectangles
        if (this.levelNumber === 3) {
            const distFromEdgeH = Math.min(col, this.brickCols - 1 - col);
            const distFromEdgeV = Math.min(row, this.brickRows - 1 - row);
            const minDist = Math.min(distFromEdgeH, distFromEdgeV);
            return minDist % 2 === 1;
        }
        
        // Level 4: Scattered pattern
        if (this.levelNumber === 4) {
            return (row * col + row + col) % 3 === 0;
        }
        
        // Level 5: Pyramid
        if (this.levelNumber === 5) {
            const midCol = Math.floor(this.brickCols / 2);
            const distFromCenter = Math.abs(col - midCol);
            return distFromCenter > row;
        }
        
        // Higher levels: More complex patterns
        // Tunnel pattern
        if (this.levelNumber === 6) {
            if (row === Math.floor(this.brickRows / 2)) {
                return true; // Leave a tunnel in the middle row
            }
            if (col === 0 || col === this.brickCols - 1 || row === 0 || row === this.brickRows - 1) {
                return false; // Always have bricks on the edges
            }
            // Random gaps
            return (row * col) % 4 === 0;
        }
        
        // Spiral pattern
        if (this.levelNumber === 7) {
            const centerRow = Math.floor(this.brickRows / 2);
            const centerCol = Math.floor(this.brickCols / 2);
            const dx = col - centerCol;
            const dy = row - centerRow;
            const distance = Math.sqrt(dx * dx + dy * dy);
            return Math.floor(distance) % 2 === 0;
        }
        
        // Higher levels use a mix of patterns with increasing difficulty
        if (this.levelNumber >= 8) {
            // More dense patterns with fewer gaps
            const complexity = Math.min(this.levelNumber - 7, 5); // Max complexity of 5
            const pattern = (row * 3 + col * 2 + this.levelNumber) % (10 - complexity);
            return pattern === 0;
        }
        
        return false;
    }
    
    /**
     * Determines the brick type based on level and position
     * @param {number} row - The row index
     * @param {number} col - The column index
     * @returns {Object} The brick type configuration object
     */
    getBrickType(row, col) {
        const brickTypes = CONFIG.BRICK_TYPES;
        
        // Level 1: All normal bricks
        if (this.levelNumber === 1) {
            return brickTypes[0];
        }
        
        // Higher levels: Mix of brick types
        // Top rows get stronger bricks
        if (row === 0) {
            // Unbreakable bricks on top row corners for higher levels
            if ((col === 0 || col === this.brickCols - 1) && this.levelNumber >= 5) {
                return brickTypes[2]; // Unbreakable
            }
            
            // Strong bricks on top row for level 3+
            if (this.levelNumber >= 3) {
                return brickTypes[1]; // Strong
            }
        }
        
        // Level 4+: Some strong bricks throughout
        if (this.levelNumber >= 4) {
            if ((row + col) % 5 === 0) {
                return brickTypes[1]; // Strong
            }
        }
        
        // Level 6+: More unbreakable bricks
        if (this.levelNumber >= 6) {
            // Unbreakable bricks in strategic positions
            if (row === 0 && col % 3 === 0) {
                return brickTypes[2]; // Unbreakable
            }
        }
        
        // Level 8+: More complex patterns with stronger bricks
        if (this.levelNumber >= 8) {
            if (row % 3 === 0 && col % 2 === 0) {
                return brickTypes[1]; // Strong
            }
            if (row === Math.floor(this.brickRows / 2) && (col === 0 || col === this.brickCols - 1)) {
                return brickTypes[2]; // Unbreakable
            }
        }
        
        // Default to normal brick
        return brickTypes[0];
    }
    
    /**
     * Determines if a brick should have an item
     * @param {number} row - The row index
     * @param {number} col - The column index
     * @returns {boolean} True if the brick should have an item, false otherwise
     */
    shouldHaveItem(row, col) {
        // Calculate a probability based on level
        // Higher levels have slightly more items
        const baseProb = 0.1;
        const levelBonus = 0.02 * Math.min(this.levelNumber, 5);
        const probability = baseProb + levelBonus;
        
        // Always put some items in specific positions for early levels
        if (this.levelNumber <= 3) {
            // Put items in middle columns
            const midCol = Math.floor(this.brickCols / 2);
            if (col === midCol || col === midCol - 1 || col === midCol + 1) {
                return Math.random() < 0.3;
            }
        }
        
        // Random chance for items
        return Math.random() < probability;
    }
    
    /**
     * Gets a random item type based on probabilities
     * @returns {string} The randomly selected item type
     */
    getRandomItemType() {
        const itemTypes = CONFIG.ITEM_TYPES;
        const totalProbability = itemTypes.reduce((sum, item) => sum + item.probability, 0);
        let random = Math.random() * totalProbability;
        
        for (const item of itemTypes) {
            if (random < item.probability) {
                return item.name;
            }
            random -= item.probability;
        }
        
        // Default to the first item type
        return itemTypes[0].name;
    }
    
    /**
     * Updates all bricks in the level
     * @param {number} deltaTime - The time passed since the last update in milliseconds
     */
    update(deltaTime) {
        for (const brick of this.bricks) {
            brick.update(deltaTime);
        }
    }
    
    /**
     * Renders all bricks in the level
     * @param {CanvasRenderingContext2D} ctx - The canvas rendering context
     */
    render(ctx) {
        for (const brick of this.bricks) {
            brick.render(ctx);
        }
    }
    
    /**
     * Handles a brick being hit
     * @param {number} index - The index of the brick in the bricks array
     * @returns {Object} Hit result {broken, points, item, brick}
     */
    hitBrick(index) {
        if (index < 0 || index >= this.bricks.length) {
            return { broken: false, points: 0, item: null, brick: null };
        }
        
        const brick = this.bricks[index];
        const result = brick.hit();
        
        if (result.broken) {
            this.brokeBricks++;
        }
        
        return {
            broken: result.broken,
            points: result.points,
            item: result.item,
            brick: brick
        };
    }
    
    /**
     * Checks if all breakable bricks are broken
     * @returns {boolean} True if all breakable bricks are broken, false otherwise
     */
    isCleared() {
        return this.brokeBricks >= this.totalBricks;
    }
    
    /**
     * Gets the completion percentage of the level
     * @returns {number} The percentage of bricks broken (0-100)
     */
    getCompletionPercentage() {
        if (this.totalBricks === 0) return 100;
        return Math.floor((this.brokeBricks / this.totalBricks) * 100);
    }
}
